File: RadiantSDK/TrueTest/Visual Studio 2013/C#/GUI Example/TrueTestEngine_CSharp_API_Example/MainForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using TrueTestEngine;
using RadiantCommon;

namespace TrueTestEngine_CSharp_API_Example
{
    public partial class MainForm : Form
    {
        public MainForm()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            var calibrationDBResult = TrueTest.APIResultEnum.Success;
            var sequenceDBResult = TrueTest.APIResultEnum.Success;
            var measurementDBResult = TrueTest.APIResultEnum.Success;

            string calibrationDatabase = Properties.Settings.Default.CalibrationDatabase;
            string sequenceFileName = Properties.Settings.Default.SequenceXMLPathFileName;
            string measurementFileName = Properties.Settings.Default.MeasurementDatabasePathFileName;

            // Initialize TrueTest
            TrueTest.Initialize(ref calibrationDatabase, 
                ref sequenceFileName, 
                ref measurementFileName, 
                this,   //Give TrueTest a reference to the main application form (this)
                ref calibrationDBResult,
                ref sequenceDBResult,
                ref measurementDBResult);

            // The strings are passed in by reference, so if the user browses to select a new file during initialization, the values will be automatically updated
            // Therefore we still need to save My.Settings.                   
            Properties.Settings.Default.CalibrationDatabase = calibrationDatabase;
            Properties.Settings.Default.SequenceXMLPathFileName = sequenceFileName;
            Properties.Settings.Default.MeasurementDatabasePathFileName = measurementFileName;
            Properties.Settings.Default.Save();

            // Add some event handlers to handle events coming from TrueTest
            AddTrueTestEventHandlers();

            // Tells TrueTest to take measurements with the camera rather than retrieving measurements from the database
            TrueTest.OperatingMode = TrueTest.OperatingModeEnum.UseCamera;

            // Makes sure to save all results into the measurement database
            TrueTest.AppSettings().SaveAnalysisResultsToDatabase = true;

            // If the user wants to make a camera measurement, make sure it gets saved to the database
            TrueTest.SaveMeasurementsToDatabase = true;

            this.Text = TrueTest.get_Sequence().Name;
        }

        private void Form1_FormClosing(object sender, EventArgs e)
        {
            TrueTest.Shutdown();
        }

        private void AddTrueTestEventHandlers()
        {
            TrueTest._MeasurementComplete += TrueTest__MeasurementComplete;
            TrueTest._MeasurementChanged += TrueTest__MeasurementChanged;
            TrueTest._DefectListChanged += TrueTest__DefectListChanged;
            TrueTest._ROIArrayChanged += TrueTest__ROIArrayChanged;
            TrueTest._AnalysisComplete += TrueTest__AnalysisComplete;
            TrueTest.SequenceComplete += TrueTest_SequenceComplete;   
        }

        private void RunButton_Click(object sender, EventArgs e)
        {
            // Disables buttons on the main form
            this.EnableButtons(false);

            // Check if we should use the camrea or database
            if (this.UseCameraCheckbox.Checked)
            {
                TrueTest.OperatingMode = TrueTest.OperatingModeEnum.UseCamera;
            }
            else
            {
                TrueTest.OperatingMode = TrueTest.OperatingModeEnum.UseDatabaseAuto;
            }

            // Clear and reset the bitmap control
            RiBitmapCtl1.ROIArray = null;
            RiBitmapCtl1.ClearDefects();
            RiBitmapCtl1.Measurement = null;
            RiBitmapCtl1.SetBackColor(Color.Black);

            // Ask for the serial number.  If the user quits or cancels, exit without starting the sequence
            string serialnumber = string.Empty;
            if (TrueTest.ShowSerialNumberDialog(ref serialnumber) != System.Windows.Forms.DialogResult.OK)
            {
                EnableButtons(true);
                return;
            }

            TrueTest.SequenceRunAll();
        }

        private void EnableButtons(bool Enabled)
        {
            if (InvokeRequired)
            {
                BeginInvoke((Action)(() => { EnableButtons(Enabled); }));
            }
            else
            {
                RunButton.Enabled = Enabled;
                EditButton.Enabled = Enabled;
                if (Enabled) RunButton.Focus();
            }
        }

        private void EditButton_Click(object sender, EventArgs e)
        {
            TrueTest.ShowEditSequenceDialog();
        }

        #region TrueTest Event Handling
        private void UpdateBitmap(MeasurementBase m)
        {
            if (InvokeRequired)
            {
                BeginInvoke((Action)(() => { UpdateBitmap(m); }));
            }
            else
            {
                RiBitmapCtl1.Measurement = m;
            }
        }

        private void SetDefectList(List<DefectBase> DefectList)
        {
            if (InvokeRequired)
            {
                BeginInvoke((Action)(() => { SetDefectList(DefectList); }));
            }
            else
            {
                RiBitmapCtl1.SetDefectList(DefectList);
                Application.DoEvents();
            }

        }

        private void SetROIArray(RegionOfInterest[] ROIArray)
        {
            if (InvokeRequired)
            {
                BeginInvoke((Action)(() => { SetROIArray(ROIArray); }));
            }
            else
            {
                RiBitmapCtl1.ROIArray = ROIArray;
                Application.DoEvents();
            }
        }

        void TrueTest_SequenceComplete(object sender, SequenceCompleteEventsArgs e)
        {
            var s = "Failed Tests= " + e.NumFailedTests.ToString();
            MessageBox.Show("Sequence Complete: " + e.SequenceName + Environment.NewLine + s);
            EnableButtons(true);

            if (e.PassFail == TrueTest.AnalysisResultEnum.Fail)
            {
                // *FAIL*
            }
        }

        void TrueTest__AnalysisComplete(object sender, string e)
        {
            var mode = ObjectRepository.ReadWriteEnum.ReadOnlyNoWait;            
            var ea = ObjectRepository.GetItem(e, ref mode) as AnalysisCompleteEventArgs;
            if (ea != null)
            {
                string s = ea.Results.Count > 0 ? ea.Results[0].Name + "= " + ea.Results[0].Value.ToString() : "";
                MessageBox.Show("Analysis Complete: " + ea.AnalysisUserName + Environment.NewLine + s + Environment.NewLine + ea.PassFail.ToString());

                if (ea.PassFail == TrueTest.AnalysisResultEnum.Fail)
                {
                    // cancel the rest of the sequence
                    TrueTest.SequenceStop();
                }
            }
        }

        void TrueTest__ROIArrayChanged(object sender, string e)
        {
            var mode = ObjectRepository.ReadWriteEnum.ReadOnlyNoWait;
            var ea = ObjectRepository.GetItem(e, ref mode) as ROIArrayChangedEventArgs;
            if (ea != null)
            {
                SetROIArray(ea.ROIArray);
            }
        }

        void TrueTest__DefectListChanged(object sender, string e)
        {
            if (!TrueTest.AppSettings().DisplayDefectList)
            {
                return;
            }

            var mode = ObjectRepository.ReadWriteEnum.ReadOnlyNoWait;
            var ea = ObjectRepository.GetItem(e, ref mode) as DefectListChangedEventArgs;
            if (ea != null)
            {
                SetDefectList(ea.DefectList);
            }
        }

        void TrueTest__MeasurementChanged(object sender, string e)
        {
            if (!TrueTest.AppSettings().DisplayUpdatedMeasurements)
            {
                return;
            }

            var mode = ObjectRepository.ReadWriteEnum.ReadOnlyNoWait;
            var ea = ObjectRepository.GetItem(e, ref mode) as MeasurementChangedEventArgs;
            if (ea != null)
            {
                UpdateBitmap(ea.Measurement);
            }
        }

        void TrueTest__MeasurementComplete(object sender, string e)
        {
            if (!TrueTest.AppSettings().DisplayCameraMeasurements)
            {
                return;
            }

            var mode = ObjectRepository.ReadWriteEnum.ReadOnlyNoWait;
            var ea = ObjectRepository.GetItem(e, ref mode) as MeasurementCompleteEventArgs;
            if (ea != null)
            {
                UpdateBitmap(ea.Measurement);
            }
        }
        #endregion
               
    }
}

File: RadiantSDK/TrueTest/Visual Studio 2013/VB.NET/GUI Example/MainForm.vb
Imports TrueTestEngine
Imports RadiantCommon
Imports System.Windows.Forms
Imports System.Drawing

Public Class MainForm

    Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
        'Initialize TrueTest
        TrueTest.Initialize(My.Settings.CalibrationDatabase, My.Settings.SequenceXMLPathFileName, My.Settings.MeasurementDatabasePathFileName, Me)
        'The strings are passed in by reference, so if the user browses to select a new file during initialization, the values will be automatically updated
        'Therefore we still need to save My.Settings.
        'The last argument is the main application form, which in most VB.NET cases is Me.  If using a language other than .NET to create GUIs, you can set this to Nothing (null).
        My.Settings.Save()

        'add some event handlers to handle events coming from TrueTest
        AddTrueTestEventHandlers()

        'Hard-code some specific settings in TrueTest for this Demo application.
        TrueTest.OperatingMode = TrueTest.OperatingModeEnum.UseDatabaseUserSelect 'pulls measurements from a database instead of measuring new with the camera
        TrueTest.AppSettings.SaveAnalysisResultsToDatabase = True 'makes sure to save all results into the measurement database
        TrueTest.SaveMeasurementsToDatabase = True 'if the user wants to make a camera measurement, make sure it gets saved to the database

        Me.Text = TrueTest.Sequence.Name
    End Sub

    Private Sub Form1_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        TrueTest.Shutdown()
    End Sub

    Private Sub AddTrueTestEventHandlers()
        AddHandler TrueTest._MeasurementComplete, AddressOf MeasurementComplete
        AddHandler TrueTest._MeasurementChanged, AddressOf MeasurementChanged
        AddHandler TrueTest._DefectListChanged, AddressOf DefectListChanged
        AddHandler TrueTest._ROIArrayChanged, AddressOf ROIArrayChanged
        AddHandler TrueTest._AnalysisComplete, AddressOf AnalysisComplete
        AddHandler TrueTest.SequenceComplete, AddressOf SequenceComplete
    End Sub

    Private Sub RunButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles RunButton.Click

        'disable buttons on the main form
        EnableButtons(False)

        'Check if we should use the camrea or database
        If (Me.UseCameraCheckbox.Checked) Then
            TrueTest.OperatingMode = TrueTest.OperatingModeEnum.UseCamera
        Else
            TrueTest.OperatingMode = TrueTest.OperatingModeEnum.UseDatabaseAuto
        End If

        'clear and reset the bitmap control
        RiBitmapCtl1.ROIArray = Nothing
        RiBitmapCtl1.ClearDefects()
        RiBitmapCtl1.Measurement = Nothing
        RiBitmapCtl1.SetBackColor(Color.Black)

        'force a redraw of the bitmap control
        Application.DoEvents()

        'ask for the serial number.  If the user quits or cancels, exit without starting the sequence
        If TrueTest.ShowSerialNumberDialog() <> System.Windows.Forms.DialogResult.OK Then
            EnableButtons(True)
            Return
        End If

        'start running the sequence
        TrueTest.SequenceRunAll()

    End Sub

    Private Sub EnableButtons(ByVal Enabled As Boolean)
        If InvokeRequired Then BeginInvoke(New Action(Sub() EnableButtons(Enabled))) : Exit Sub
        RunButton.Enabled = Enabled
        EditButton.Enabled = Enabled
        If Enabled Then RunButton.Focus()
    End Sub

    Private Sub EditButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles EditButton.Click

        'shows a form to edit the sequence
        TrueTest.ShowEditSequenceDialog()

    End Sub

#Region "TrueTest Event Handling"

    Public Sub MeasurementComplete(ByVal sender As Object, ByVal e As String)
        If TrueTest.AppSettings.DisplayCameraMeasurements = False Then
            Return
        End If

        Dim obj As Object = ObjectRepository.GetItem(e, ObjectRepository.ReadWriteEnum.ReadOnlyNoWait)
        If obj IsNot Nothing Then
            Dim ea As TrueTestEngine.MeasurementCompleteEventArgs = CType(obj, TrueTestEngine.MeasurementCompleteEventArgs)
            UpdateBitmap(ea.Measurement)
        End If
    End Sub

    Private Sub MeasurementChanged(ByVal sender As Object, ByVal e As String)
        If TrueTest.AppSettings.DisplayUpdatedMeasurements = False Then
            Return
        End If

        Dim obj As Object = ObjectRepository.GetItem(e, ObjectRepository.ReadWriteEnum.ReadOnlyNoWait)
        If obj IsNot Nothing Then
            Dim ea As TrueTestEngine.MeasurementChangedEventArgs = CType(obj, TrueTestEngine.MeasurementChangedEventArgs)
            UpdateBitmap(ea.Measurement)
        End If
    End Sub

    Private Sub UpdateBitmap(ByVal m As MeasurementBase)
        If InvokeRequired Then BeginInvoke(New Action(Sub() UpdateBitmap(m))) : Exit Sub

        Me.RiBitmapCtl1.Measurement = m
    End Sub

    Private Sub DefectListChanged(ByVal sender As Object, ByVal e As String)
        If TrueTest.AppSettings.DisplayDefectList = False Then
            Return
        End If

        Dim obj As Object = ObjectRepository.GetItem(e, ObjectRepository.ReadWriteEnum.ReadOnlyNoWait)
        If obj IsNot Nothing Then
            Dim ea As TrueTestEngine.DefectListChangedEventArgs = CType(obj, TrueTestEngine.DefectListChangedEventArgs)
            SetDefectList(ea.DefectList)
        End If
    End Sub

    Friend Sub SetDefectList(ByVal DefectList As List(Of DefectBase))
        Me.RiBitmapCtl1.SetDefectList(DefectList)
        Application.DoEvents()
    End Sub

    Private Sub ROIArrayChanged(ByVal sender As Object, ByVal e As String)
        Dim obj As Object = ObjectRepository.GetItem(e, ObjectRepository.ReadWriteEnum.ReadOnlyNoWait)
        If obj IsNot Nothing Then
            Dim ea As TrueTestEngine.ROIArrayChangedEventArgs = CType(obj, TrueTestEngine.ROIArrayChangedEventArgs)
            SetROIArray(ea.ROIArray)
        End If
    End Sub

    Friend Sub SetROIArray(ByVal ROIArray() As RegionOfInterest)
        Me.RiBitmapCtl1.ROIArray = ROIArray
        Application.DoEvents()
    End Sub

    Private Sub SequenceComplete(ByVal sender As Object, ByVal e As TrueTestEngine.SequenceCompleteEventsArgs)
        Dim s As String = "Failed Tests= " & e.NumFailedTests.ToString
        MessageBox.Show("Sequence Complete: " & e.SequenceName & vbCrLf & s)
        EnableButtons(True)

        If e.PassFail = TrueTest.AnalysisResultEnum.Fail Then
            '*FAIL*
        End If
    End Sub

    Private Sub AnalysisComplete(ByVal sender As Object, ByVal e As String)
        Dim obj As Object = ObjectRepository.GetItem(e, ObjectRepository.ReadWriteEnum.ReadOnlyNoWait)
        If obj IsNot Nothing Then
            Dim ea As TrueTestEngine.AnalysisCompleteEventArgs = CType(obj, TrueTestEngine.AnalysisCompleteEventArgs)
            If ea.Results.Count > 0 Then
                Dim s As String = ea.Results(0).Name & "= " & ea.Results(0).Value.ToString
                MessageBox.Show("Analysis Complete: " & ea.AnalysisUserName & vbCrLf & s & vbCrLf & ea.PassFail.ToString)
            End If

            If ea.PassFail = TrueTest.AnalysisResultEnum.Fail Then

                    'cancel the rest of the sequence
                    TrueTest.SequenceStop()

                End If
            End If
    End Sub
#End Region
End Class